source/change-streams/tests/README.rst

```diff
@@ -0,0 +1,156 @@
+
+- Else
+
+  - Assert that ``actual`` is of the same JSON type as ``expected``
+  - If ``expected`` is a ``number``, ``string``, ``boolean`` or ``null``
+
+    - Assert that ``expected`` equals ``actual``
+
+  - Else if ``expected`` is an ``array``
+
+    - for every ``idx``/``value`` in ``expected``
+
+      - Assert that ``value`` MATCHES ``actual[idx]``
+
+  - Else
+
+    - for every ``key``/``value`` in ``expected``
+
+      - Assert that ``value`` MATCHES ``actual``
+
+Spec Test Runner
+================
+
+Before running the tests
+
+- Create a MongoDB server topology.
+- Create a MongoClient ``globalClient``, and connect to the server
+
+For each YAML file, for each element in ``tests``:
+
+- If ``topology`` does not include the topology of the server instance(s), skip this test.
+- Use ``globalClient`` to
+
+  - Drop the database ``database_name``
+  - Create the database ``database_name`` and the collection ``database_name.collection_name``
+
+- Create a new MongoClient ``client``, with APM enabled
+- Using ``client``, create a changeStream ``changeStream`` against the specified ``target``. Use ``changeStreamPipeline`` and ``changeStreamOptions`` if they are non-empty
+- Using ``globalClient``, run every operation in ``operations`` in serial against the server
+- Wait until either
+
+  - An error occurres
+  - All operations have been successful AND the changeStream has received as many changes as there are ``result.success``
+
+- Close ``changeStream``
+- If there was an error
+
+  - Assert that an error was expected for the test.
+  - Assert that the error MATCHES ``results.error``
+
+- Else
+
+  - Assert that no error was expected for the test
+  - Assert that the changes received from ``changeStream`` MATCH the results in ``results.success``
+
+- If there are any ``expectations``
+
+  - For each APM ``commandStarted`` events on ``client``
+
+    - Filter out any ``ismaster`` events
+
+  - For each (``expected``, ``idx``) in ``expectations``
+
+    - Assert that ``actual[idx]`` MATCHES ``expected``
+
+- Close the MongoClient ``client``
+
+After running all tests
+
+- Close the MongoClient ``globalClient``
+- Shut down the server topology
+
+
+Prose Tests
+===========
+
+The following tests have not yet been automated, but MUST still be tested
+
+1. ``ChangeStream`` must continuously track the last seen ``resumeToken``
+2. ``ChangeStream`` will throw an exception if the server response is missing the resume token
+3. ``ChangeStream`` will automatically resume one time on a resumable error (including `not master`) with the initial pipeline and options, except for the addition/update of a ``resumeToken``.
+4. ``ChangeStream`` will not attempt to resume on a server error
+5. ``ChangeStream`` will perform server selection before attempting to resume, using initial ``readPreference``
+6. Ensure that a cursor returned from an aggregate command with a cursor id, and an initial empty batch, is not closed on the driver side.
```

Comment:
I don't see any need to wrap "and an initial empty batch" in a comma-delimited clause.